File: solutions/c/pascals-triangle/1/pascals_triangle.c
#include "pascals_triangle.h"

#include <malloc.h>
#include <string.h>

#define ZERO2ONE(nrows) ( (nrows>0) ? nrows : 1)

void fill_row(uint8_t**, size_t);

void fill_row(uint8_t **triangle, size_t i){
        for (size_t j = 0; j<(i+1); j++){
            if (0==j || i == j) 
                triangle[i][j] = 1;
            else 
                triangle[i][j] = triangle[i-1][j-1] + triangle[i-1][j];     
        }
}

uint8_t **create_triangle(size_t rows)
{
    uint8_t **triangle = calloc(ZERO2ONE(rows), sizeof(uint8_t*));
    if(!triangle)  return NULL;
    if(0==rows){
        triangle[0] = calloc(1, sizeof(triangle[0]));
        if (!triangle[0]) {free(triangle); return NULL;}
        return triangle;
    }
    for(size_t i = 0;i<rows; i++){
        triangle[i] = calloc(rows ,sizeof(uint8_t));
        memset(triangle[i], 0, rows*sizeof(uint8_t));
        fill_row(triangle, i);
    }
    return triangle;
}

void free_triangle(uint8_t **triangle, size_t rows){
    if(!rows)
        return;
     for(size_t i = 0;i<ZERO2ONE(rows); i++){
        free(triangle[i]);
    }
    free(triangle);
    return;
}

File: solutions/dart/pascals-triangle/1/lib/pascals_triangle.dart
class PascalsTriangle {
  List<List<int>> rows(int n){
    if (n<=0)
      return <List<int>>[];
    List<List<int>> pascal = [<int>[1]];
    for(var r=1; r<n; r++){
      List<int> current = [];
      for(var k=0; k<=r; k++){
        int rkCoeff;
        if(k==0 || k==r)
          rkCoeff = 1;
        else
          rkCoeff = pascal[r-1][k-1] + pascal[r-1][k];
        current.add(rkCoeff);
      }
      pascal.add(current);
    }
    return pascal;
  }
}
